### PR TITLE
Fix bug in AC input power calculation

### PR DESCRIPTION
**Pull request by DanielMcInnes**
Specific to AC inputs on Multi RS systems. Tested OK on RS and non-RS systems.

### CODE REVIEW

File: data/common/ActiveAcInput.qml [hunk 1]
```diff
@@ -27,25 +27,15 @@ Device {
...
-						total = Units.sumRealNumbers(total, phase ? phase[propertyName] : NaN)
-					}
-				}
-				totalPower = total
-			} else if (propertyName === "current" && count === 1) {
-				firstPhaseCurrent = propertyValue
+		readonly property int totalPower: {
+			let sum = 0
+			for (let i = 0; i < count; ++i) {
+				sum += get(i).power || 0
```
Comment (DanielMcInnes) [inline data/common/ActiveAcInput.qml:33]: The individual phase power values are integers, so there is no need to use 'Units.sumRealNumbers'

Comment (blammit) [reply data/common/ActiveAcInput.qml:33]: Can you move this calculation to the `multiComponent` in AcInputServiceLoader? The total doesn't need to be calculated for grid/genset/vebus as those services provide total power values. We only need to do it for Multi RS.

Comment (blammit) [reply data/common/ActiveAcInput.qml:33]: Well I guess it's harder to move there directly due to the model here, but can you make sure this is only calculated for the multi case?

Comment (blammit) [reply data/common/ActiveAcInput.qml:33]: Or, we could use a timer like gui-v1 to always calculate it manually, and avoid fetching the overall power from AcInputServiceLoader for grid/genset/vebus.

**Review by blammit — COMMENTED** [review]
Could you remove the `_totalPower` and `power` properties/calculations from AcInputServiceLoader as well? I'm pretty sure this means they are no longer used.

File: data/common/ActiveAcInput.qml (hunk 1 above)
Comment (DanielMcInnes) [reply data/common/ActiveAcInput.qml]: updated as suggested

**Comment by DanielMcInnes** [comment]
> Could you remove the `_totalPower` and `power` properties/calculations from AcInputServiceLoader as well? I'm pretty sure this means they are no longer used.

done

**Comment by chriadam** [comment]
Looks good, although I'm not a huge fan of the fixed interval timer.  Could you improve the commit message so that it explains what the original bug was, and why/how the commit fixes it?

**Comment by blammit** [comment]
>> Could you remove the _totalPower and power properties/calculations from AcInputServiceLoader as well? I'm pretty sure this means they are no longer used.
> done

@DanielMcInnes there is still some stale code there, can you remove the rest as well? We don't need any power calculations in AcInputServiceLoader anymore, since ActiveAcInput doesn't use them anymore. See:

- `hasTotalPower` property
- vebusComponent: power, _powerValue, _power VeQuickItem
- gridOrGensetComponent: power, _power

File: data/common/AcInputPhaseModel.qml [hunk 2]
```diff
@@ -9,8 +9,6 @@ import Victron.VenusOS
 ListModel {
 	id: root
 
-	signal phaseValueChanged(int phaseIndex, string propertyName, real propertyValue)
-
 	readonly property bool _feedbackEnabled: Global.systemSettings.essFeedbackToGridEnabled
```
Comment (blammit) [inline data/common/AcInputPhaseModel.qml:12]: Is this unused now? And same for the _feedbackEnabled checks elsewhere in this file. If unused they can also be removed.

**Comment by DanielMcInnes** [comment]
> Looks good, although I'm not a huge fan of the fixed interval timer. Could you improve the commit message so that it explains what the original bug was, and why/how the commit fixes it?

Done. Could you have a look at the commit message please, and let me know if you're happy with it?

File: data/common/AcInputPhaseModel.qml (hunk 2 above)
Comment (DanielMcInnes) [reply data/common/AcInputPhaseModel.qml:12]: '_feedbackEnabled' is still used

**Comment by DanielMcInnes** [comment]
> > > Could you remove the _totalPower and power properties/calculations from AcInputServiceLoader as well? I'm pretty sure this means they are no longer used.
> > > done
> 
> @DanielMcInnes there is still some stale code there, can you remove the rest as well? We don't need any power calculations in AcInputServiceLoader anymore, since ActiveAcInput doesn't use them anymore. See:
> 
>     * `hasTotalPower` property
> 
>     * vebusComponent: power, _powerValue, _power VeQuickItem
> 
>     * gridOrGensetComponent: power, _power

Done

**Comment by blammit** [comment]
@DanielMcInnes Regarding the commit message:

- "Fix bug in AC input power calculation" is quite generic -- we refer to AC input power from a number of places. Something like "Fix calculation of total AC input power" would be better as it's more specific.

- I don't think the code sample in the commit message is necessary

- "Now, we iterate over the phases correctly." -> but that's not all? There is also a Timer now, so it should explain why the Timer is added (i.e. to get parity with gui-v1, and have less frequent updates to the UI)

- Can you include the issue number with "Fixes #xx"

File: data/common/ActiveAcInput.qml [hunk 3]
```diff
@@ -27,23 +27,19 @@ Device {
...
+		property real firstPhaseCurrent: count === 1 ? get(0).current : NaN
 
-		onPhaseValueChanged: function(phaseIndex, propertyName, propertyValue) {
-			if (propertyName === "power" && !root._acInputService.hasTotalPower) {
-				let total = NaN
-				for (let i = 0; i < phases.count; ++i) {
-					if (i === phaseIndex) {
-						total = Units.sumRealNumbers(total, propertyValue)
-					} else {
-						const phase = get(phaseIndex)
```
Comment (DanielMcInnes) [inline data/common/ActiveAcInput.qml:40]: This is where the bug was.
`const phase = get(phaseIndex)`
should be:
`const phase = get(i)`

**Comment by DanielMcInnes** [comment]
> @DanielMcInnes Regarding the commit message:
> 
>     * "Fix bug in AC input power calculation" is quite generic -- we refer to AC input power from a number of places. Something like "Fix calculation of total AC input power" would be better as it's more specific.
> 
>     * I don't think the code sample in the commit message is necessary
> 
>     * "Now, we iterate over the phases correctly." -> but that's not all? There is also a Timer now, so it should explain why the Timer is added (i.e. to get parity with gui-v1, and have less frequent updates to the UI)
> 
>     * Can you include the issue number with "Fixes #xx"

Thanks Bea, done

**Review by blammit — APPROVED** [review]

**Pull request merged**